enhancements/machine-api/autoscaling-from-to-zero.md

```diff
@@ -0,0 +1,127 @@
+
+- As a dev I want each provider specific logic to be self contained in a single place.
+
+This proposes to let each actuator to contain the provider specific logic to fetch and expose in a machineSet (with zero replicas) the details expected by the autoscaler to be able to create a "virtual" node and scale from zero.
+
+### Implementation Details
+
+#### Machine API Autoscaler provider
+
+The autoscaler exposes [TemplateNodeInfo](https://github.com/openshift/kubernetes-autoscaler/blob/253ee49441750815c70b606f242eb76164d9bdc4/cluster-autoscaler/cloudprovider/cloud_provider.go#L152) on the `NodeGroup` interface to let provider implementers to expose the information of an empty (as if just started) node. This is used in scale-up simulations to predict what would a new node look like if a node group was expanded.
+
+- The [machine API autoscaler provider](https://github.com/openshift/kubernetes-autoscaler/tree/master/cluster-autoscaler/cloudprovider/openshiftmachineapi) `nodegroup` implementation operates over scalable resources i.e `machineSets`.
+
+- We want to enable any provider to expose in a machineSet any the details needed by the autoscaler to scale from zero. 
+
+- The autoscaler can then use this details to populate a `schedulernodeinfo.NodeInfo` object enabling scaling from zero. To this end we propose the following annotations for a machineSet:
+
+```
+machine.openshift.io/vCPU
+machine.openshift.io/memoryMb
+machine.openshift.io/GPU
+```
+
+- Details which are non provider specific can be inferred directly from the [machineSet machine template](https://github.com/openshift/machine-api-operator/blob/master/pkg/apis/machine/v1beta1/machineset_types.go#L80) e.g lables, taints.
+
+- The `TemplateNodeInfo` implementation will look very similar to [AWS](https://github.com/openshift/kubernetes-autoscaler/blob/253ee49441750815c70b606f242eb76164d9bdc4/cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go#L321)/[Azure](https://github.com/openshift/kubernetes-autoscaler/blob/253ee49441750815c70b606f242eb76164d9bdc4/cluster-autoscaler/cloudprovider/azure/azure_scale_set.go#L598) autoscaler implementations:
+but  `buildNodeFromTemplate` will get what it needs from the given `nodeGroup`/`machineSet` with zero replicas.
+
+#### Providers implementation
+
+- A new controller is developed in each actuator and run by the machine controller manager.
+
+- It watches machineSets and annotate them.
+
+- Is a per provider choice to decide how to infer the details to be included in the well known annotations.
+
+- For AWS/Azure/GCP the details can be inferred from the instance type set in `machineSet.template.Spec.ProviderSpec`. This can be done dynamically or statically [reusing the autoscaler specific provider implementation](https://github.com/openshift/kubernetes-autoscaler/blob/253ee49441750815c70b606f242eb76164d9bdc4/cluster-autoscaler/cloudprovider/aws/ec2_instance_types.go#L30).
```

Comment:
Is there some sort of API that can be used to look up CPU/Memory etc for node types within different cloud providers? If not, how would we work out this information if we have never seen an instance type in the cluster?

Reply:
> Is there some sort of API that can be used to look up CPU/Memory etc for node types within different cloud providers? If not, how would we work out this information if we have never seen an instance type in the cluster?

This can be found/inferred/queried against a specific cloud provider based on the info in the MachineSet's `ProviderSpec`.